source/core_http_client.c

```diff
@@ -2091,51 +2177,34 @@ HTTPStatus_t HTTPClient_Send( const TransportInterface_t * pTransport,
         LogError( ( "Parameter check failed: reqBodyBufLen > INT32_MAX."
                     "reqBodyBufLen=%lu",
                     ( unsigned long ) reqBodyBufLen ) );
-        returnStatus = HTTPInvalidParameter;
     }
     else
     {
-        /* Empty else for MISRA 15.7 compliance. */
+        if( pResponse->getTime == NULL )
+        {
+            /* Set a zero timestamp function when the application did not configure
+             * one. */
+            pResponse->getTime = getZeroTimestampMs;
+        }
+
+        returnStatus = HTTPSuccess;
     }
 
-    /* Send the headers, which are at one location in memory. */
     if( returnStatus == HTTPSuccess )
     {
-        returnStatus = sendHttpHeaders( pTransport,
+        returnStatus = sendHttpRequest( pTransport,
+                                        pResponse->getTime,
                                         pRequestHeaders,
+                                        pRequestBodyBuf,
                                         reqBodyBufLen,
                                         sendFlags );
     }
 
-    /* Send the body, which is at another location in memory. */
-    if( returnStatus == HTTPSuccess )
-    {
-        if( pRequestBodyBuf != NULL )
-        {
-            returnStatus = sendHttpBody( pTransport,
-                                         pRequestBodyBuf,
-                                         reqBodyBufLen );
-        }
-        else
-        {
-            LogDebug( ( "A request body was not sent: pRequestBodyBuf is NULL." ) );
-        }
-    }
-
     if( returnStatus == HTTPSuccess )
     {
-        /* If the application chooses to receive a response, then pResponse
-         * will not be NULL. */
-        if( pResponse != NULL )
-        {
-            returnStatus = receiveAndParseHttpResponse( pTransport,
-                                                        pResponse,
-                                                        pRequestHeaders );
-        }
-        else
-        {
-            LogDebug( ( "Response ignored: pResponse is NULL." ) );
-        }
+        returnStatus = receiveAndParseHttpResponse( pTransport,
+                                                    pResponse,
+                                                    pRequestHeaders );
```

Comment:
I think we still need to check if `if( pResponse != NULL )` either here or in the private method. It may break the shim if receiving the response is not optional.

Reply:
it is required now otherwise it will break the config option

Reply:
I realized we should always have the response otherwise the connection is hosed with extra data coming in.

Reply:
The old shim will always return -1 for transport recv so when we get here, we will return immediately, and the network receive task will still be able to receive the data. However, an error will be printed and returned. Should we simply not update the submodule pointer in P4?

Reply:
We may be able to update the shim to check the status code, but I'm not quite sure how to get around the logs.

Reply:
Okay let's just update the shim to return the number of bytes received.

Reply:
Ok